### PR TITLE
Localization refactor

### PR DESCRIPTION
**Pull request by pelletjl**
Refactor mana cost localization

### CODE REVIEW

File: Necrosis/XML/Attributes.lua [hunk 1]
```diff
@@ -226,11 +234,15 @@ function Necrosis:PetSpellAttribute()
 	for i = 2, 6, 1 do
 		local f = _G["NecrosisPetMenu"..i]
 		if f then
+			local SpellName_Rank = self.Spell[i+1].Name
+			if self.Spell[i+1].Rank and not (self.Spell[i+1].Rank == " ") then
+				SpellName_Rank = SpellName_Rank.."("..self.Spell[i+1].Rank..")"
+			end
```
Comment (tekHudson) [inline Necrosis/XML/Attributes.lua]: These 4 lines appear multiple times. let's abstract this into a helper method.

File: Necrosis/XML/Attributes.lua [hunk 2]
```diff
@@ -241,8 +253,12 @@ function Necrosis:PetSpellAttribute()
 	for i = 1, #buttonID, 1 do
 		local f = _G["NecrosisPetMenu"..buttonID[i]]
 		if f then
+			local SpellName_Rank = self.Spell[ BuffID[i] ].Name
+			if self.Spell[ BuffID[i] ].Rank and not (self.Spell[ BuffID[i] ].Rank == " ") then
+				SpellName_Rank = SpellName_Rank.."("..self.Spell[ BuffID[i] ].Rank..")"
+			end
```
Comment (tekHudson) [inline Necrosis/XML/Attributes.lua]: These 4 lines appear multiple times. let's abstract this into a helper method.

File: Necrosis/XML/Attributes.lua [hunk 3]
```diff
@@ -257,10 +273,14 @@ function Necrosis:CurseSpellAttribute()
 	for i = 1, #buffID, 1 do
 		local f = _G["NecrosisCurseMenu"..i]
 		if f then
+			local SpellName_Rank = self.Spell[ buffID[i] ].Name
+			if self.Spell[ buffID[i] ].Rank and not (self.Spell[ buffID[i] ].Rank == " ") then
+				SpellName_Rank = SpellName_Rank.."("..self.Spell[ buffID[i] ].Rank..")"
+			end
```
Comment (tekHudson) [inline Necrosis/XML/Attributes.lua]: These 4 lines appear multiple times. let's abstract this into a helper method.

File: Necrosis/XML/Attributes.lua (hunk 1 above)
Comment (tekHudson) [reply Necrosis/XML/Attributes.lua]:
```
function Necrosis:SpellNameWithRank(spellId)
    local SpellName_Rank = self.Spell[spellId].Name
    if self.Spell[i+1].Rank and not (self.Spell[spellId].Rank == " ") then
        SpellName_Rank = SpellName_Rank.."("..self.Spell[spellId].Rank..")"
    end
    SpellName_Rank 
end

local SpellName_Rank = SpellNameWithRank(i+1)

Comment (tekHudson) [reply Necrosis/XML/Attributes.lua]: it looks like this new method could be used in other place, I will comment on them in this PR.

File: Necrosis/XML/Attributes.lua [hunk 4]
```diff
@@ -151,10 +151,14 @@ function Necrosis:BuffSpellAttribute()
 	-- Association de l'armure demoniaque si le sort est disponible
 	if _G["NecrosisBuffMenu1"] then
 		NecrosisBuffMenu1:SetAttribute("type", "spell")
-		if not self.Spells[31].spellId then
-			NecrosisBuffMenu1:SetAttribute("spell", self.Spells[36].spellId)
+		if not self.Spell[31].ID then
+			NecrosisBuffMenu1:SetAttribute("spell",
+				self.Spell[36].Name.."("..self.Spell[36].Rank..")"
```
Comment (tekHudson) [inline Necrosis/XML/Attributes.lua]: Use new method discussed below here.

File: Necrosis/XML/Attributes.lua [hunk 5]
```diff
@@ -151,10 +151,14 @@ function Necrosis:BuffSpellAttribute()
...
+			)
 		else
-			NecrosisBuffMenu1:SetAttribute("spell", self.Spells[31].spellId)
+			NecrosisBuffMenu1:SetAttribute("spell",
+				self.Spell[31].Name.."("..self.Spell[31].Rank..")"
```
Comment (tekHudson) [inline Necrosis/XML/Attributes.lua]: Use new method discussed below here.

File: Necrosis/XML/Attributes.lua [hunk 6]
```diff
@@ -169,19 +173,23 @@ function Necrosis:BuffSpellAttribute()
 			if not (i == 2 or i == 5 or i == 7 or i == 8 or i == 9 or i == 10) then
 				f:SetAttribute("unit", "target")
 			end
-			f:SetAttribute("spell", self.Spells[buffID[i]].spellId)
+			local SpellName_Rank = self.Spell[ buffID[i] ].Name
+			if self.Spell[ buffID[i] ].Rank and not (self.Spell[ buffID[i] ].Rank == " ") then
+				SpellName_Rank = SpellName_Rank.."("..self.Spell[ buffID[i] ].Rank..")"
+			end
```
Comment (tekHudson) [inline Necrosis/XML/Attributes.lua]: These 4 lines appear multiple times. let's abstract this into a helper method.

File: Necrosis/XML/Attributes.lua [hunk 7]
```diff
@@ -312,11 +332,18 @@ function Necrosis:StoneAttribute(Steed)
 	NecrosisSpellTimerButton:SetAttribute("macrotext", "/focus")
 	NecrosisSpellTimerButton:SetAttribute("type2", "item")
 	NecrosisSpellTimerButton:SetAttribute("item", self.Translation.Item.Hearthstone)
+
+	-- if the 'Ritual of Souls' spell is known, then associate it to the hearthstone shift-click.
```
Comment (tekHudson) [inline Necrosis/XML/Attributes.lua]: `Ritual of Souls` isn't a part of classic, Doesn't exist in spell ID chart. Suggest removing

File: Necrosis/Initialize.lua [hunk 8]
```diff
@@ -99,7 +99,6 @@ function Necrosis:Initialize(Config)
 		self.Spell[index].ID = nil
 	end
 	self:SpellLocalize()
-	self:CreateSpellList()
```
Comment (tekHudson) [inline Necrosis/Initialize.lua]: Where is the replacement for this? I see yuou removed the object creation below, what is populating `self.Spell`?

File: Necrosis/Initialize.lua [hunk 9]
```diff
@@ -167,7 +166,6 @@ function Necrosis:Initialize(Config)
 		self.Speech.Rez = self.Speech.ShortMessage[1]
 		self.Speech.TP = self.Speech.ShortMessage[2]
 	end
-	self:Msg("Ready", "USER")
```
Comment (tekHudson) [inline Necrosis/Initialize.lua]: 👍

File: Necrosis/Initialize.lua (hunk 8 above)
Comment (tekHudson) [reply Necrosis/Initialize.lua]: Nevermind, I found it's creation in `function Necrosis:SpellSetup()`

Nice

File: Necrosis/Necrosis.lua [hunk 10]
```diff
@@ -1626,7 +1506,7 @@ function Necrosis:UpdateMana()
 	-- If shadow guardian cooldown we gray || Si cooldown de gardien de l'ombre on grise
 	if _G["NecrosisBuffMenu8"] and self.Spell[43].ID then
 		local start, duration = GetSpellCooldown(self.Spell[43].ID, "spell")
-		if self.Spells[43].Mana > mana and start > 0 and duration > 0 then
+		if self.Spell[43].Mana > mana and start > 0 and duration > 0 then
```
Comment (tekHudson) [inline Necrosis/Necrosis.lua]: All these mana calculations will not work I believe. The mana values will not always reflect actual cost. All spells have a mana cost of 50 in the object and we retrieve the basic spell info. See Healthstone for instance:
```
[52] = {Name = GetSpellInfo(5699),	Mana = 50,				Length = 0,		Type = 0}, -- Create Healthstone
```
The ID referenced here is for https://classic.wowhead.com/spell=5699/create-healthstone
What if character is higher level and needs to cast https://classic.wowhead.com/spell=11729/create-healthstone-greater?

File: Necrosis/Spells.lua [hunk 11]
```diff
@@ -258,96 +258,11 @@ function Necrosis:SpellLocalize(tooltip)
...
+		GameTooltip:AddLine(self.Spell[spellIndex].Mana.." "..self.Translation.Misc.Mana)
 	end
 
-	for i = 1, #self.SpellRef, 1 do
-		self.Spells[i] = {}
-		for index, value in pairs(self.SpellRef[i]) do
-			if IsSpellKnown(value) then
-				local Name = GetSpellInfo(value)
-				self.Spells[i].Name = Name
-				self.Spells[i].Mana = self:GetManaCost(value)
```
Comment (tekHudson) [inline Necrosis/Spells.lua]: we were updating mana cost here.

File: Necrosis/Spells.lua [hunk 12]
```diff
@@ -258,96 +258,11 @@ function Necrosis:SpellLocalize(tooltip)
...
-			[54] = {17953,17952,17951,6366},								-- Create Firestone || Création pierre de feu
-			[55] = {18938,18937,18220},										-- Dark Pact || Pacte noir
-			[56] = {},														-- NOPE NOT IN Classic  Shadow Cleave || Enchainement d'ombre
-			[57] = {},														-- NOPE NOT IN Classic  Immolation Aura || Aura d'immolation
-			[58] = {},														-- NOPE NOT IN Classic  Challenging Howl || Hurlement de défi
-			[59] = {}														-- NOPE NOT IN Classic   Demonic Empowerment || Renforcement démoniaque
-		}
+function Necrosis:ManaCostLocalize(spellIndex)
+	if GetLocale() == "ruRU" then
+		GameTooltip:AddLine(self.Translation.Misc.Mana..": "..self.Spell[spellIndex].Mana)
```
Comment (tekHudson) [inline Necrosis/Spells.lua]: Unless we are updating mana values somewhere we only set to 50 in the object creation.

**Comment by pelletjl** [comment]
the first commit in this chain is just a giant revert of the branch that was merged in from cerka-refacqto branch. i was trying to minimize the changes, but it was difficult because of the commits afterward. I'll see what i can do about cleaning it up. there's some good ideas in there but it seemed to bite off a bit too much and ended up breaking some stuff along the way.

**Comment by tekHudson** [comment]
Thanks so much for your hard work. We'll get this thing all squared away!

**Pull request merged**